### PR TITLE
Bump Typescript to v4

### LINKED ISSUE #1784: Bump Typescript to 4.1
Should we?

----

There are some type issues to fix in some packages after quickly checking

### PR DESCRIPTION
**Pull request by dapplion**
Fixes type breaking changes to be compatible with Typecript v4, so we can benefit from the performance improvements and new features.

See https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/
Fixes https://github.com/ChainSafe/lodestar/issues/1784

### CODE REVIEW

**Comment by dapplion** [comment]
A lot of unit tests are failing, anyone has a clue what might have changed to cause so many failures?

File: packages/lodestar-utils/src/bytes.ts [hunk 1]
```diff
@@ -35,14 +34,6 @@ export function bytesToBigInt(value: Uint8Array, endianness: Endianness = "le"):
   throw new Error("endianness must be either 'le' or 'be'");
 }
 
-export function toHex(buffer: ArrayLike<number>): string {
```
Comment (mpetrunic) [inline packages/lodestar-utils/src/bytes.ts:38]: Can you put this method back. We should use this method instead of Buffer.from().toHex().

This change doesn't belong in this PR anyways

**Review by mpetrunic — COMMENTED** [review]
Only breaking changes that could cause tests to fail are:
- https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#getterssetters-are-no-longer-enumerable
- https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#parsing-differences-in-optional-chaining-and-non-null-assertions

You should probably debug one tests and detect line thats causing trouble

**Comment by wemeetagain** [comment]
Seems the stubs are no longer working

Eg:
in lodestar-beacon-state-transition/test/unit/stateTransition/epoch/justification.test.ts
set breakpoint on L43
getCurrentEpochStub.called === false

Seems related to the getters no longer being enumerable?
https://github.com/microsoft/TypeScript/issues/38568
@mpetrunic any thoughts?

**Comment by mpetrunic** [comment]
> Seems the stubs are no longer working
> 
> Eg:
> in lodestar-beacon-state-transition/test/unit/stateTransition/epoch/justification.test.ts
> set breakpoint on L43
> getCurrentEpochStub.called === false
> 
> Seems related to the getters no longer being enumerable?
> [microsoft/TypeScript#38568](https://github.com/microsoft/TypeScript/issues/38568)
> @mpetrunic any thoughts?

Typical response from Ryan...
There is nice tread on sinon repo https://github.com/sinonjs/sinon/issues/2258

I think it worked before because typescript didn't generated esm compliant code (aka module imports were mutable). If my memory serves me correctly, babel was already doing that so we switched from babel-register to ts-node/register.

Few ideas:
- try this for running tests https://github.com/jasmine/jasmine/issues/1817#issuecomment-630110087
- switch to jest - it hacks into module loading system and allows stubbing but I think it also caches loaded modules
- switch back to babel-register with plugin like this that would transpile modules to be mutable for test only https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs

**Comment by dapplion** [comment]
We could use an import mocking library such as https://www.npmjs.com/package/rewiremock

I've used it with success and offers full type support. Example:

```ts
  const { default: getPortsToOpen } = await rewiremock.around(
      () => import("../../../src/watchers/natRenewal/getPortsToOpen"),
      mock => {
        mock(() => import("../../../src/modules/docker/listContainers"))
          .with({ listContainers })
          .toBeUsed();
      }
    );
```

From https://github.com/dappnode/DNP_DAPPMANAGER/blob/cbd7cbcf3e155f0dd7e9e969ee37b431fb916819/packages/dappmanager/test/watchers/natRenewal/getPortsToOpen.test.ts#L74

**Comment by mpetrunic** [comment]
> We could use an import mocking library such as https://www.npmjs.com/package/rewiremock
> 
> I've used it with success and offers full type support. Example:
> 
> ```ts
>   const { default: getPortsToOpen } = await rewiremock.around(
>       () => import("../../../src/watchers/natRenewal/getPortsToOpen"),
>       mock => {
>         mock(() => import("../../../src/modules/docker/listContainers"))
>           .with({ listContainers })
>           .toBeUsed();
>       }
>     );
> ```
> 
> From https://github.com/dappnode/DNP_DAPPMANAGER/blob/cbd7cbcf3e155f0dd7e9e969ee37b431fb916819/packages/dappmanager/test/watchers/natRenewal/getPortsToOpen.test.ts#L74
> 

Im pretty sure that only works because you are using typescript <3.9. All those libs were just modifying exports. Feel free to try it though

**Comment by dapplion** [comment]
> Im pretty sure that only works because you are using typescript <3.9. All those libs were just modifying exports. Feel free to try it though

RIght, I have not tried it with Typescript >= 3.9

**Comment by dapplion** [comment]
rewiremock works fine with Typescript v4.
- I've skipped all tests in the beacon state transition that were mocking imports
- I've refactored the tests in lodestar that were mocking imports with rewiremock. It's not pretty, and slow but allows us to move forward.

All our code does dependency injection everywhere, where we pass entire instances of the `chain`, `network`, etc. However, we don't do dependency injection for BLS and state transition utils. To fix this type of issues with mocking imports that result in very ugly verbose test code we could also do dependency injection for these remaining items

**Review by wemeetagain — APPROVED** [review]
This is good for now to move forward.
I think ideally, we can build out better testing infrastructure that can help us test on valid-ish states so we can avoid relying on mocking and just have better e2e tests.
